subsystems/viamserver/viamserver.go

```diff
@@ -294,17 +297,80 @@ func (s *viamServer) HealthCheck(ctx context.Context) (errRet error) {
 	return errRet
 }
 
+func (s *viamServer) isRestartAllowed(ctx context.Context) (restartAllowed bool, errRet error) {
+	for _, url := range []string{s.checkURL, s.checkURLAlt} {
+		s.logger.Debugf("starting restart allowed check for %s using %s", SubsysName, url)
+
+		ctx, cancelFunc := context.WithTimeout(ctx, time.Second*10)
+		defer cancelFunc()
+
+		url += "/restart_status"
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+		if err != nil {
+			return false, errw.Wrapf(err, "checking whether %s allows restart", SubsysName)
+		}
+
+		// disabling the cert verification because it doesn't work in offline mode (when connecting to localhost)
+		//nolint:gosec
+		client := &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}}
+
+		resp, err := client.Do(req)
+		if err != nil {
+			return false, errw.Wrapf(err, "checking whether %s allows restart", SubsysName)
+		}
+
+		defer func() {
+			utils.UncheckedError(resp.Body.Close())
+		}()
+
+		var restartStatusResponse rdkweb.RestartStatusResponse
+		err = json.NewDecoder(resp.Body).Decode(&restartStatusResponse)
+		if err != nil {
+			// Only debug-log errors from JSON decoding. Interacting with older
+			// viam-server instances will result in a decoding error, as empty HTML
+			// will be returned.
+			s.logger.Debugf("ignoring error from restart allowed check for %s: %v",
+				SubsysName, err.Error())
+		} else {
+			return restartStatusResponse.RestartAllowed, nil
+		}
+	}
+	return false, nil
+}
+
 func (s *viamServer) Update(ctx context.Context, cfg *pb.DeviceSubsystemConfig, newVersion bool) (bool, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	setFastStart(cfg)
+
+	// By default, return false on the needRestart flag, as we await the user to
+	// kill/restart viam-server directly.
+	var needRestart bool
+
 	if newVersion && s.running {
-		s.logger.Info("awaiting user restart to run new viam-server version")
 		s.shouldRun = false
+
+		// viam-server can be safely restarted even while running if the process
+		// has reported it is safe to do so through its `restart_status` HTTP
+		// endpoint.
+		restartAllowed, err := s.isRestartAllowed(ctx)
+		if err != nil {
+			return false, err
```

Comment:
```suggestion
			return needRestart, err
```
nit: but if you're gonna have a variable set for this, should use it

Reply:
Done.